allensdk/brain_observatory/behavior/behavior_ophys_session.py

```diff
@@ -305,6 +306,45 @@ def segmentation_mask_image(self) -> Image:
     def segmentation_mask_image(self, value):
         self._segmentation_mask_image = value
 
+    @property
+    def eye_tracking_data(self) -> pd.DataFrame:
+        """A dataframe containing ellipse fit parameters for the eye, pupil
+        and corneal reflection (cr). Fits are derived from tracking points
+        from a DeepLabCut model applied to video frames of a subject's
+        right eye. Raw tracking points and raw video frames are not exposed
+        by the SDK.
+
+        Notes:
+        - All columns starting with 'pupil_' represent ellipse fit parameters
+          relating to the pupil.
+        - All columns starting with 'eye_' represent ellipse fit parameters
+          relating to the eyelid.
+        - All columns starting with 'cr_' represent ellipse fit parameters
+          relating to the corneal reflection, which is caused by an infrared
+          LED positioned near the eye tracking camera.
+        - All positions are in units of pixels.
+        - All areas are in units of pixels^2
+        - All values are in the coordinate space of the eye tracking camera,
+          NOT the coordinate space of the stimulus display (i.e. this is not
+          gaze location), with (0, 0) being the upper-left corner of the
+          eye-tracking image.
+        - The 'likely_blink' column is True for any row (frame) where the pupil
+          fit failed OR eye fit failed OR an outlier fit was identified.
+        - All ellipse fits are derived from tracking points that were output by
+          a DeepLabCut model that was trained on hand-annotated data frome a
+          subset of imaging sessions on optical physiology rigs.
+        - Raw DeepLabCut tracking points are not publicly available.
+
+        :rtype: pandas.DataFrame
+        """
+        if self._eye_tracking_data is None:
+            self._eye_tracking_data = self.api.get_eye_tracking_data()
+        return self._eye_tracking_data
+
+    @eye_tracking_data.setter
+    def eye_tracking_data(self, value):
```

Comment:
You have some internal methods that have configuration variables (e.g. `determine_outliers` has `z_threshold`, and `determine_likely_blinks` has `dilation_frames`). I can see a use-case where the user would want to have control over these variables, but there's no way to actually change them from the defaults either by configuring the API or through the session. You should include a way to set them from the session somewhere.

Reply:
Good idea, do you think changing `eye_tracking_data` from a property to a method like `get_eye_tracking_data` for the behavior+ophys session object would be a reasonable way to expose those parameters?

Reply:
I'd like to keep the access pattern the same as what's already been done so that it's not confusing for users. I think maybe instead we could add these arguments as optional parameters when the behavior session is initialized -- either as keyword arguments or as a dictionary of keyword arguments. They would be stored as session attributes. We'd update the logging to provide users with logs at the `info` level to clarify which arguments were used when `eye_tracking_data` is accessed via the session object.

Reply:
The only part that might muddy the waters about this is that if users update the attributes, they won't re-pull the eye tracking data since it'll already be cached. We'll need to decide how to deal with this (require users to create a new session object since technically everything should be read-only? use get/set logic on the session config vars to know when to "refresh" the data? etc.)

Reply:
I don't have any opinions/ideas on the implementation details, but I do agree with @kschelonka that it would be ideal for users to be able to control these parameters.

Reply:
I think since we don't necessarily want to re-pull all the data just from changing these configs, I suggest that we set these as properties, and use the setter method to set the value for `self._eye_tracking_data` to `None`. That will trigger pulling that data (and that data only) again. Include enough logging to make it clear to users what is happening.

Reply:
@kschelonka @dougollerenshaw Okay, I've updated this PR to add the described feature. Let me know if any other changes are desired!

Reply:
@njmei: Either the parameter change isn't working, or I'm not using it properly. Following I used your notebook, then added a cell to plot the likely_blink column vs. time for two different settings of dilation_frames. But the result didn't change. Here's the code:

```
fig,ax=plt.subplots()
session.eye_tracking_processing_args = {"z_threshold": 1.0, "dilation_frames": 2}
eye_data1 = session.eye_tracking_data.copy(deep=True)
ax.plot(
    eye_data1['time'],
    eye_data1['likely_blink'],
    color = 'black'
)

session.eye_tracking_processing_args = {"z_threshold": 1.0, "dilation_frames": 4}
eye_data2 = session.eye_tracking_data.copy(deep=True)
ax.plot(
    eye_data2['time'],
    eye_data2['likely_blink'],
    color = 'red',
    linewidth=4,
    alpha=0.5,
    linestyle=':'
)

ax.legend(['dilation frames = 2', 'dilation frames = 4'])
ax.set_xlim(2576,2581)
```

And here's the result (I'd expect the red-dashed line to rise sooner and fall later relative to the black line, but instead they are identical)
![image](https://user-images.githubusercontent.com/19944442/76805452-0b474280-679c-11ea-9a85-1fdc86ea6f2e.png)

I'm getting the notification that the underlying parameter is changing, so it looks like something is happening in the background:
![image](https://user-images.githubusercontent.com/19944442/76805559-595c4600-679c-11ea-8995-93aa2ca40680.png)

Reply:
@dougollerenshaw Doh! I forgot to commit a change, so wasted time not being able to reproduce your example. I've updated the examples notebook to demonstrate the correct behavior.

Reply:
Thanks @njmei! I'm seeing a couple weird things now.
1) When I run the exact same code block as above, the black line (for dilation_frames=2) is different than before. Why would that one have changed?

Here's the plot:
![image](https://user-images.githubusercontent.com/19944442/76810979-3b4b1180-67ad-11ea-8cf7-f3177ef09b54.png)

Reply:
2) If I set dilation_frames = 0, then *every* frame gets labeled as a likely blink:
![image](https://user-images.githubusercontent.com/19944442/76811062-75b4ae80-67ad-11ea-86e6-6823f38cf25e.png)
![image](https://user-images.githubusercontent.com/19944442/76811101-92e97d00-67ad-11ea-8ac8-545fd3174a26.png)

Reply:
3) If I set dilation_frames=1, it's not clear to me what's triggering the likely_blink column to be set to True. For example, here's an example where indices 20231 to 20233 are labeled likely_blink=True. I'd expect the middle value (index 20232) to have a NaN or outlier fit. It doesn't:
![image](https://user-images.githubusercontent.com/19944442/76811279-376bbf00-67ae-11ea-9734-2edb4feda95e.png)
What did the algorithm identify that caused it to label these three frames as likely_blink=True?

Reply:
@dougollerenshaw 

1. Because of the commit I forgot, setting `eye_tracking_processing_args` was not actually doing anything and you were likely seeing the results of eye tracking data processed with defaults: `{"z_threshold": 3.0, "dilation_frames": 2}`

2. Because dilation is accomplished with the `scipy.ndimage.binary_dilation` function the documentation states that: [iterations int, optional

    The dilation is repeated iterations times (one, by default). If iterations is less than 1, the dilation is repeated until the result does not change anymore. Only an integer of iterations is accepted.
](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_dilation.html)

I can change this behavior so that a `dilation_frame=0` does no `binary_dilation` if desired.

3. If any of `cr_area`, `eye_area`, or `pupil_area` are outliers then that would cause them to be labeled as a `likely_blink`.

Reply:
Thanks @njmei!

1) You're right, when I changed back to `z_threshold=3`, everything acts as expected (I added dot markers to make it clear that the red line rises two frames earlier and falls two frames later than the black line, as expected):
![image](https://user-images.githubusercontent.com/19944442/76812145-35572f80-67b1-11ea-87ba-5c91fb5bb504.png)

2) Thanks for digging into the documentation. That's very weird behavior for `n_iterations=0`! I agree with your proposal: If `dilation_frames=0`, do not call `binary_dilation`.

3) Yes, when I set `z_threshold=3`, that made `likely_blink=False` for index 20232 (and its neighbors). So you were likely right that it was the corneal reflection fit that triggered that, given the low threshold (`z_threshold=1`) for outliers.